File: uikit/src/main/java/io/agora/chat/uikit/chat/viewholder/EaseVoiceViewHolder.java
package io.agora.chat.uikit.chat.viewholder;

import android.media.MediaPlayer;
import android.os.AsyncTask;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;


import java.io.File;


import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.VoiceMessageBody;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.interfaces.MessageListItemClickListener;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowVoice;
import io.agora.chat.uikit.widget.chatrow.EaseChatRowVoicePlayer;
import io.agora.exceptions.ChatException;
import io.agora.util.EMLog;

public class EaseVoiceViewHolder extends EaseChatRowViewHolder{
    private EaseChatRowVoicePlayer voicePlayer;

    public EaseVoiceViewHolder(@NonNull View itemView, MessageListItemClickListener itemClickListener) {
        super(itemView, itemClickListener);
        voicePlayer = EaseChatRowVoicePlayer.getInstance(getContext());
    }

    @Override
    public void onBubbleClick(ChatMessage message) {
        super.onBubbleClick(message);
        String msgId = message.getMsgId();

        if (voicePlayer.isPlaying()) {
            // Stop the voice play first, no matter the playing voice item is this or others.
            voicePlayer.stop();
            // Stop the voice play animation.
            ((EaseChatRowVoice) getChatRow()).stopVoicePlayAnimation();

            // If the playing voice item is this item, only need stop play.
            String playingId = voicePlayer.getCurrentPlayingId();
            if (msgId.equals(playingId)) {
                return;
            }
        }

        if (message.direct() == ChatMessage.Direct.SEND) {
            // Play the voice
            String localPath = ((VoiceMessageBody) message.getBody()).getLocalUrl();
            File file = new File(localPath);
            if (file.exists() && file.isFile()) {
                playVoice(message);
                // Start the voice play animation.
                ((EaseChatRowVoice) getChatRow()).startVoicePlayAnimation();
            } else {
                asyncDownloadVoice(message);
            }
        } else {
            final String st = getContext().getResources().getString(R.string.ease_is_download_voice_click_later);
            if (message.status() == ChatMessage.Status.SUCCESS) {
                if (ChatClient.getInstance().getOptions().getAutodownloadThumbnail()) {
                    play(message);
                } else {
                    VoiceMessageBody voiceBody = (VoiceMessageBody) message.getBody();
                    EMLog.i("TAG", "Voice body download status: " + voiceBody.downloadStatus());
                    switch (voiceBody.downloadStatus()) {
                        case PENDING:// Download not begin
                        case FAILED:// Download failed
                            getChatRow().updateView(message);
                            asyncDownloadVoice(message);
                            break;
                        case DOWNLOADING:// During downloading
                            Toast.makeText(getContext(), st, Toast.LENGTH_SHORT).show();
                            break;
                        case SUCCESSED:// Download success
                            play(message);
                            break;
                    }
                }
            } else if (message.status() == ChatMessage.Status.INPROGRESS) {
                Toast.makeText(getContext(), st, Toast.LENGTH_SHORT).show();
            } else if (message.status() == ChatMessage.Status.FAIL) {
                Toast.makeText(getContext(), st, Toast.LENGTH_SHORT).show();
                asyncDownloadVoice(message);
            }
        }
    }

    private void playVoice(ChatMessage msg) {
        voicePlayer.play(msg, new MediaPlayer.OnCompletionListener() {
            @Override
            public void onCompletion(MediaPlayer mp) {
                // Stop the voice play animation.
                ((EaseChatRowVoice) getChatRow()).stopVoicePlayAnimation();
            }
        });
    }

    private void asyncDownloadVoice(final ChatMessage message) {
        new AsyncTask<Void, Void, Void>() {
            @Override
            protected Void doInBackground(Void... params) {
                ChatClient.getInstance().chatManager().downloadAttachment(message);
                return null;
            }

            @Override
            protected void onPostExecute(Void result) {
                super.onPostExecute(result);
                getChatRow().updateView(message);
            }
        }.execute();
    }

    private void play(ChatMessage message) {
        String localPath = ((VoiceMessageBody) message.getBody()).getLocalUrl();
        File file = new File(localPath);
        if (file.exists() && file.isFile()) {
            ackMessage(message);
            playVoice(message);
            // Start the voice play animation.
            ((EaseChatRowVoice) getChatRow()).startVoicePlayAnimation();
        } else {
            EMLog.e("TAG", "file not exist");
        }
    }

    private void ackMessage(ChatMessage message) {
        ChatMessage.ChatType chatType = message.getChatType();
        if (!message.isAcked() && chatType == ChatMessage.ChatType.Chat) {
            try {
                ChatClient.getInstance().chatManager().ackMessageRead(message.getFrom(), message.getMsgId());
            } catch (ChatException e) {
                e.printStackTrace();
            }
        }
        if (!message.isListened()) {
            ChatClient.getInstance().chatManager().setVoiceMessageListened(message);
        }
    }

    @Override
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        if(voicePlayer.isPlaying()) {
            voicePlayer.stop();
            voicePlayer.release();
        }
    }
}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/feature/chat/viewholders/EaseVoiceViewHolder.kt
package io.agora.uikit.feature.chat.viewholders

import android.os.AsyncTask
import android.view.View
import android.widget.Toast
import io.agora.uikit.R
import io.agora.uikit.common.ChatClient
import io.agora.uikit.common.ChatDownloadStatus
import io.agora.uikit.common.ChatException
import io.agora.uikit.common.ChatMessage
import io.agora.uikit.common.ChatMessageDirection
import io.agora.uikit.common.ChatMessageStatus
import io.agora.uikit.common.ChatVoiceMessageBody
import io.agora.uikit.common.helper.EaseChatRowVoicePlayer
import io.agora.uikit.widget.chatrow.EaseChatRowVoice
import io.agora.uikit.common.ChatLog
import io.agora.uikit.common.ChatType
import java.io.File

class EaseVoiceViewHolder(itemView: View) : EaseChatRowViewHolder(itemView) {
    private val voicePlayer: EaseChatRowVoicePlayer = EaseChatRowVoicePlayer.getInstance(getContext())

    override fun onBubbleClick(message: ChatMessage?) {
        super.onBubbleClick(message)
        message?.run {
            val msgId: String = msgId
            if (voicePlayer.isPlaying) {
                // Stop the voice play first, no matter the playing voice item is this or others.
                voicePlayer.stop()
                // Stop the voice play animation.
                (getChatRow() as EaseChatRowVoice?)?.stopVoicePlayAnimation()

                // If the playing voice item is this item, only need stop play.
                val playingId: String? = voicePlayer.currentPlayingId
                if (msgId == playingId) {
                    return
                }
            }
            if (direct() === ChatMessageDirection.SEND) {
                // Play the voice
                val localPath: String = (body as ChatVoiceMessageBody).localUrl
                val file = File(localPath)
                if (file.exists() && file.isFile) {
                    playVoice(this)
                    // Start the voice play animation.
                    (getChatRow() as EaseChatRowVoice?)?.startVoicePlayAnimation()
                } else {
                    asyncDownloadVoice(this)
                }
            } else {
                val st = getContext().resources.getString(R.string.ease_is_download_voice_click_later)
                val voiceBody = body as ChatVoiceMessageBody
                val downloadStatus = voiceBody.downloadStatus()
                if (status() === ChatMessageStatus.SUCCESS) {
                    if (ChatClient.getInstance().options.autodownloadThumbnail && downloadStatus == ChatDownloadStatus.SUCCESSED) {
                        play(this)
                    } else {
                        ChatLog.i("TAG", "Voice body download status: $downloadStatus")
                        when (downloadStatus) {
                            ChatDownloadStatus.PENDING, ChatDownloadStatus.FAILED -> {
                                getChatRow()?.updateView()
                                asyncDownloadVoice(this)
                            }

                            ChatDownloadStatus.DOWNLOADING -> Toast.makeText(getContext(), st, Toast.LENGTH_SHORT).show()
                            ChatDownloadStatus.SUCCESSED -> play(this)
                        }
                    }
                } else if (status() === ChatMessageStatus.INPROGRESS) {
                    Toast.makeText(getContext(), st, Toast.LENGTH_SHORT).show()
                } else if (status() === ChatMessageStatus.FAIL) {
                    Toast.makeText(getContext(), st, Toast.LENGTH_SHORT).show()
                    asyncDownloadVoice(this)
                } else {

                }
            }
        }

    }

    private fun playVoice(msg: ChatMessage) {
        voicePlayer.play(msg) { // Stop the voice play animation.
            (getChatRow() as EaseChatRowVoice?)?.stopVoicePlayAnimation()
        }
    }

    private fun asyncDownloadVoice(message: ChatMessage) {
        object : AsyncTask<Void?, Void?, Void?>() {
            override fun doInBackground(vararg params: Void?): Void? {
                ChatClient.getInstance().chatManager().downloadAttachment(message)
                return null
            }

            override fun onPostExecute(result: Void?) {
                super.onPostExecute(result)
                getChatRow()?.updateView()
            }
        }.execute()
    }

    private fun play(message: ChatMessage) {
        val localPath: String = (message.body as ChatVoiceMessageBody).localUrl
        val file = File(localPath)
        if (file.exists() && file.isFile) {
            ackMessage(message)
            playVoice(message)
            // Start the voice play animation.
            (getChatRow() as EaseChatRowVoice?)?.startVoicePlayAnimation()
        } else {
            ChatLog.e("TAG", "file not exist")
        }
    }

    private fun ackMessage(message: ChatMessage) {
        val chatType = message.chatType
        if (!message.isAcked && chatType === ChatType.Chat) {
            try {
                ChatClient.getInstance().chatManager()
                    .ackMessageRead(message.from, message.msgId)
            } catch (e: ChatException) {
                e.printStackTrace()
            }
        }
        if (!message.isListened) {
            ChatClient.getInstance().chatManager().setVoiceMessageListened(message)
        }
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        if (voicePlayer.isPlaying) {
            voicePlayer.stop()
            voicePlayer.release()
        }
    }
}